[lldb/swift] Add swift type-system progress reporting

medismailben opened the pull request:
This patch reports progress events specific to the swift type-system.

Currently, it covers 3 long-running operations:
- Setting up the runtime reflection context
- Loading implicit modules from the compile unit
- Caching them in the expression persistent state

It also includes a test a will listen for swift progress events and make
sure it we receive them when fetching a swift variable.

rdar://75803015

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

medismailben commented:
@swift-ci test

medismailben commented:
@swift-ci test macOS platform

kastiglione commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -3712,8 +3713,16 @@ void SwiftASTContext::LoadModule(swift::ModuleDecl *swift_module,
+                    imported_modules.size());
+
+  size_t completion = 0;
+  for (auto import : imported_modules) {
```
while you're here, this looks like it could be `const auto &import`

kastiglione commented on lldb/test/API/functionalities/progress_reporting/swift_progress_reporting/TestSwiftProgressReporting.py:
```diff
@@ -0,0 +1,67 @@
+
+    mydir = TestBase.compute_mydir(__file__)
+
+    eBroadcastBitStopProgressThread = (1 << 0)
```
is this temporary & should it be declared in the SB API?

medismailben replied:
This is specific to the test, it's just used to the tell the `fetch_events` loop we're done listening to new events.

kastiglione replied:
thanks, I guess the name threw me off. Maybe it doesn't need the `e` prefix since it's not from an enum (that is what the `e` stands for right?)

kastiglione commented on lldb/test/API/functionalities/progress_reporting/swift_progress_reporting/TestSwiftProgressReporting.py:
```diff
@@ -0,0 +1,67 @@
+                    if event_mask & self.eBroadcastBitStopProgressThread:
+                        done = True;
+                elif event.BroadcasterMatchesRef(self.progress_broadcaster):
+                    ret_args = lldb.SBDebugger().GetProgressFromEvent(event);
```
should `GetProgressFromEvent` be a static function?

medismailben replied:
It is static internally. To unpack the event data, we don't need to have access to a specific instance of the debugger.

JDevlieghere replied:
Seems like we're starting to accumulate a few of these functions. Can we put this in a helper function upstream that's generic enough to be used by this tests as well as the two upstream ones?

medismailben replied:
@JDevlieghere I'll do that in a follow-up patch

kastiglione commented on lldb/test/API/functionalities/progress_reporting/swift_progress_reporting/TestSwiftProgressReporting.py:
```diff
@@ -0,0 +1,67 @@
+        self.test_broadcaster.BroadcastEventByType(self.eBroadcastBitStopProgressThread)
+        listener_thread.join()
+
+        self.assertGreater(len(self.progress_events), 0)
```
should the assertions on `progress_events` be more fine grained?

JDevlieghere commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -3948,12 +3957,21 @@ void SwiftASTContext::ValidateSectionModules(
     if (!GetModule(module_info, error))
       module.ReportWarning("unable to load swift module \"%s\" (%s)",
                            module_name.c_str(), error.AsCString());
+    else
```
By putting this in the else branch, we might report less events than `module_names.size()`. Why not increment the value even if the module couldn't be loaded?

medismailben replied:
Good call!

JDevlieghere commented on lldb/test/API/functionalities/progress_reporting/swift_progress_reporting/main.swift:
```diff
@@ -0,0 +1,49 @@
+struct IntPair {
```
Was this copied from another test? If so, can we use the original source file instead of duplicating it? It doesn't seem like the content of this tests is particularly relevant for the test. I think it's fine to have a path that goes outside the current directory in the Makefile.

medismailben replied:
This was copied from the `SwiftUnsafe` data formatter test (that I wrote few years ago). I chose it because it creates various types of variables, that will - hopefully - exercise the type-system to report events. However, I modified it a bit because there were some bits that are not relevant to this test (i.e. turning an array into a `UnsafePointer` / `UnsafeBufferPointer` etc.).

@adrian-prantl may be you have some suggestions for a simpler test ?

adrian-prantl replied:
You'll want to use a test that imports other modules. The test I added to https://github.com/apple/llvm-project/pull/4083/files has a pretty minimal example for adding a framework (you can remove the hidden second framework and you also don't need to use a .in file for the source code of the dylib).

medismailben replied:
@adrian-prantl I updated the test in https://github.com/apple/llvm-project/pull/4085/commits/d1bfff86e80dafd512c7bd2dd49ee79a4851bc28

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntime.cpp:
```diff
@@ -461,36 +462,51 @@ void SwiftLanguageRuntimeImpl::SetupReflection() {
+  if (!exe_module) {
+    LLDB_LOGF(GetLog(LLDBLog::Types), "%s: Failed to get executable module",
+              LLVM_PRETTY_FUNCTION);
+    m_initialized_reflection_ctx = false;
```
This is a change in behavior, but I think it's fine.

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntime.cpp:
```diff
@@ -461,36 +462,51 @@ void SwiftLanguageRuntimeImpl::SetupReflection() {
 
+  Progress progress(llvm::formatv("Setting up swift reflection for {0}",
+                                  exe_module->GetObjectName().AsCString()),
+                    m_modules_to_add.GetSize());
```
Nice.

adrian-prantl commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -3712,8 +3713,16 @@ void SwiftASTContext::LoadModule(swift::ModuleDecl *swift_module,
+  size_t completion = 0;
+  for (auto import : imported_modules) {
     import.importedModule->collectLinkLibraries(addLinkLibrary);
+    progress.Increment(++completion);
```
I think this one is too deep inside. What do think about instead moving this progress meter to
SwiftASTContext::CacheUserImports() and SwiftASTContext::GetCompileUnitImportsImpl() which are calling into this function?

adrian-prantl replied:
By "to deep" I mean that this loop happens *after* the module has already been loaded and is comparatively quick.
What happens if a Progress created while another Progress is already in flight? If that's handled gracefully, we can keep this too and just add the outer loop.

medismailben commented:
@swift-ci test

medismailben committed:
[lldb/test] Refine TestSwiftProgressReporting (NFC)

This patch simplifies the inferior for TestSwiftProgressReporting, by
creating a variable using a type defined in an imported module.

It also improve the test by making sure it receives each kind of
expected swift progress events.

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

medismailben commented:
@swift-ci test linux platform

medismailben commented:
@swift-ci test macOS platform

medismailben committed:
[lldb/test] Add events listener helper class to lldbtest

This patch introduces a generic helper class that will listen for
event in a background thread and match it against a source broadcaster.

If the event received matches the source broadcaster, the event is
queued up in a list that the user can access later on.

The motivation behind this is to easily test new kinds of events
(i.e. Swift type-system progress events). However, this patch also
updates `TestProgressReporting.py` and `TestDiagnosticReporting.py`
to make use of this new helper class.

Differential Revision: https://reviews.llvm.org/D121977

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

medismailben committed:
[lldb/bindings] Update SWIG static bindings (NFC)

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

medismailben committed:
[lldb/swift] Add swift type-system progress reporting

This patch reports progress events specific to the swift type-system.

Currently, it covers 3 long-running operations:
- Setting up the runtime reflection context
- Loading implicit modules from the compile unit
- Caching them in the expression persistent state

It also includes a test a will listen for swift progress events and make
sure it we receive them when fetching a swift variable.

rdar://75803015

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

medismailben commented:
@swift-ci test

dduan commented:
`TestSwiftProgressReporting.py` seems to be sporadically failing on apple/swift's main branch. See examples [1](https://ci.swift.org/job/swift-PR-macos-smoke-test/757/console) and [2](https://ci.swift.org/job/swift-PR-macos-smoke-test/753/console)

medismailben commented:
There is a race happening only in the test which cause it to fail randomly. I'm working on a fix on llvm.org, will cherry-pick the fix afterwards.

Since this only concerns the test, I don't think we should revert all the patches.

medismailben commented:
This discussion is happening on Phab, should be landing the fix today or tomorrow https://reviews.llvm.org/D122193

medismailben commented:
Should be fix with #4109
